### ISSUE TITLE
mobx-react: computed properties aren't recomputed while accessing React context with static contextType

### ISSUE THREAD

**Issue by pixelkritzel** [mobxjs/mobx#3392]
This is a CodeSandbox displaying the bug.

https://codesandbox.io/s/zealous-tesla-wioeys?file=/src/Item.tsx

The checkboxes don't update.

This version works because the context field is made observable. But this causes warnings in the console.

https://codesandbox.io/s/sweet-wilbur-2m70bd?file=/src/Item.tsx

**Comment by urugator** [comment]
Guess we would have to make `this.context` implicitly observable similary to `this.props` and `this.state`:
https://github.com/mobxjs/mobx/blob/2e14aad3e283bf75041e4bfc956ddc9ce5b6580b/packages/mobx-react/src/observerClass.ts#L46-L51
@mweststrate opinions?

Hackish workaround
```javascript
@observable
_context!: FormContextType;
set context(val: FormContextType){
  runInAction(() => this._context = val)    
};
get context() {
  return this._context;
}
```

You can also avoid relying on context propagation and pass something stable+observable that children can subscribe to (eg Store instance itself).

**Comment by pixelkritzel** [comment]
@mweststrate 

My assumption was this would to the same trick. But obviously it doesn't work.

```
@observable
  selectedId: FormContextType["selectedId"];

  @action
  setSelectedId = (id: string) => {
    console.log("setting value");
    this.selectedId = id;
  };

  @computed
  get contextValue(): FormContextType {
    return { selectedId: this.selectedId, setSelectedId: this.setSelectedId };
  }
  
 ```

**Comment by mweststrate** [comment]
That is because you copy selectedId into a new non obsevable object. It's
the property you made observable,  but the value you're passing down
(primitive values dont change themselved)

**Comment by mweststrate** [comment]
https://mobx.js.org/understanding-reactivity.html for some more context. But note that this doesn't really relate to MobX, but just to how JavaScript works :) What you did is roughly the equivalent to:

```
const store = {
  selectedId: 1
}

context = {
   selectedId: store.selectedId // this copies the current value into a new prop, doesn't alias!
}

printSelectedId() {
  console.log(store.selectedId)
}

printSelectedId() // prints 1

store.selectedId = 2

printSelectedId() // still prints 1, not 2! context and store are unrelated objects

```

**Comment by urugator** [comment]
@mweststrate 
>  But note that this doesn't really relate to MobX

I think that's not completely true. If you remove the `@computed` from the getter, therefore taking Mobx out of the picture, everything works. The problem here is that computed caches the value, that depens on context, but it doesn't invalidate together with the component, when it recieves new context...

**Comment by pixelkritzel** [comment]
Also it doesn't work to wrap the returned object of the computed property in `observable`.

https://codesandbox.io/s/keen-khorana-uqcmul?file=/src/Form.tsx

The problem is the subscribing component doesn't update.

**Comment by urugator** [comment]
@pixelkritzel This is expected, as mentioned you would have to pass something observable + stable ... meaning the context value itself never changes - it must be still the same object reference with an observable property `selectedId`.

**Comment by pixelkritzel** [comment]
> Guess we would have to make this.context implicitly observable similary to this.props and this.state:

For me this seems to have the most consistent outcome in a way of an observed component to behave like a React component. But I'm not sure what could be all the unintended consequences.

**Comment by pixelkritzel** [comment]
> The problem with unstable classic context is it can absolutely kill your application (they solved that with useContext, which is subscription based), since passing a new context value in the classic context api way will rerender your full tree. See [medium.com/@mweststrate/how-to-safely-use-react-context-b7e343eff076](https://medium.com/@mweststrate/how-to-safely-use-react-context-b7e343eff076)

This references the old legacy Context API while my use case uses the current stable Context API.

**Comment by pixelkritzel** [comment]
> Ermöglichen ein Icon links neben text anzuzeigen

Does this mean `this.context` could be observable by default? Or is it possible to exclude certain properties from the strict mode check?

**Comment by urugator** [comment]
`this.context` is only relevant when `Class.contextType` is set, correct? Just considering possible optimizations...

**Comment by pixelkritzel** [comment]
> this.context is only relevant when Class.contextType is set, correct? Just considering possible optimizations...

Yes. Because it is assigned automatically by React **(if Class.contextType is set)** strict mode will flood the console with warnings.

https://reactjs.org/docs/context.html#classcontexttype

**Comment by pixelkritzel** [comment]
@urugator Do you change this or should I open a pull request? (It would be pretty simple, making this.context observable if `contextType` is not falsy.)

**Comment by urugator** [comment]
I am quite busy these days, so PR is welcome. But please don't forget to provide tests as well.
I am ready to accept this change, but I dunno about @mweststrate, so please keep that in mind.

**Comment by pixelkritzel** [comment]
> But please don't forget to provide tests as well.

is this normal?

This is the output after 

```
$ npm install
$ npm build
$ npm build:test
$ npm test
```

![Screenshot 2022-05-18 at 23 12 27](https://user-images.githubusercontent.com/1437379/169156747-76123f53-3b6f-47c5-801e-002500f1f5bd.png)

**Comment by urugator** [comment]
Nope, use `yarn`

**Comment by pixelkritzel** [comment]
Is there some documentation how to successfully building the package and running the tests?

![Screenshot 2022-05-18 at 23 19 54](https://user-images.githubusercontent.com/1437379/169157719-a0ea9d97-96a8-4c70-80c0-2b2b00268c6c.png)

**Comment by urugator** [comment]
You probably forget to delete `node_modules` after that `npm ...`

Following should work (just tried):
```javascript
git clone git@github.com:mobxjs/mobx.git
cd mobx
yarn install
yarn lerna run build
yarn test
```
=>
```
Test Suites: 78 passed, 78 total
Tests:       31 skipped, 1301 passed, 1332 total
Snapshots:   111 passed, 111 total
Time:        62.995 s
Ran all test suites in 5 projects.
Done in 64.58s.
```

**Comment by pixelkritzel** [comment]
@urugator Thx!

**Comment by pixelkritzel** [comment]
https://github.com/mobxjs/mobx/pull/3404

**Comment by urugator** [comment]
Fixed in mobx-react@7.5.0
